### PR TITLE
[SYCL][E2E] Fix some e2e tests memory leak.

### PR DESCRIPTION
**Pull request by yingcong-wu**
Fix some leaks found in e2e tests.

### CODE REVIEW

**Review by kbenzie — COMMENTED** [review]
Replacing these raw pointer allocations with `std::vector` would be more idiomatic C++, then they would be freed automatically. Is there a reason not to do that?

**Comment by yingcong-wu** [comment]
> Replacing these raw pointer allocations with `std::vector` would be more idiomatic C++, then they would be freed automatically. Is there a reason not to do that?

Good advice! No particular reason not to do that, I just went for the minimum change to fix the problems. I will change to using vector now.

**Comment by turinevgeny** [comment]
LGTM for ESIMD.

**Review by turinevgeny — APPROVED** [review]

File: sycl/test-e2e/ESIMD/histogram_256_slm.cpp [hunk 1]
```diff
@@ -207,5 +207,7 @@ int main() {
   else
     std::cout << "FAILED\n";
 
+  free(hist);
```
Comment (v-klochkov) [inline sycl/test-e2e/ESIMD/histogram_256_slm.cpp:210]: Minor comment: 'hist' is allocated with 'new'[], not 'malloc', thus being pedantic it should be deallocated with 'delete[] hist;' not 'free(hist)'.

Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/histogram_256_slm.cpp:210]: Ok, free() may work for int* allocated with new[] as `int` type does not need calling destructor for every element

Comment (yingcong-wu) [reply sycl/test-e2e/ESIMD/histogram_256_slm.cpp]: Fixed.

**Review by v-klochkov — APPROVED** [review]

**Review by kbenzie — APPROVED** [review]
UR LGTM

**Pull request merged**